baselines/fedmlb/README.md

```diff
@@ -0,0 +1,602 @@
+| $\lambda_2$  | Used in FedMLB. See Eq. 8 in the original paper. It weights the impact of hybrid KL loss.| 1.0 |
+| temperature  | Regulates the smoothness of local and global predictions in distillation. | 1.0 |
+| $\gamma$   | Used in FedAvg+KD. It weights the impact of local-global distillation on local training.| 0.2 |
+
+## Environment Setup
+By default, Poetry will use the Python version in your system. 
+In some settings, you might want to specify a particular version of Python 
+to use inside your Poetry environment. You can do so with `pyenv`. 
+Check the documentation for the different ways of installing `pyenv`,
+but one easy way is using the automatic installer:
+
+```bash
+curl https://pyenv.run | bash
+```
+You can then install any Python version with `pyenv install <python-version>`
+(e.g. `pyenv install 3.9.17`) and set that version as the one to be used. 
+```bash
+# cd to your FedMLB directory (i.e. where the `pyproject.toml` is)
+pyenv install 3.10.6
+
+pyenv local 3.10.6
+
+# set that version for poetry
+poetry env use 3.10.6
+```
+To build the Python environment as specified in the `pyproject.toml`, use the following commands:
+```bash
+# cd to your FedMLB directory (i.e. where the `pyproject.toml` is)
+
+# install the base Poetry environment
+poetry install
+
+# activate the environment
+poetry shell
+```
+## Running the Experiments
+Ensure you have activated your Poetry environment (execute `poetry shell` from
+this directory).
+
+### Generating clients' dataset
+First (and just the first time), the data partitions of clients must be generated.
+
+#### CIFAR-100 clients' dataset generation
+
+To generate the partitions for the CIFAR-100 settings, i.e.:
+1. Moderate-scale with Dir(0.3), 100 clients, balanced dataset (500 examples per client) -- **default config**.
+2. Large-scale experiments with Dir(0.3), 500 clients, balanced dataset (100 examples per client);
+3. Moderate-scale with Dir(0.6), 100 clients, balanced dataset (500 examples per client);
+4. Large-scale experiments with Dir(0.6), 500 clients, balanced dataset (100 examples per client),
+
+use the following commands:
+```bash
+# this will run using the default settings in the `conf/base.yaml`
+# and will generate the setting for 1. (see above)
+python -m fedmlb.dataset_preparation 
+
+# this will generate the setting for 2. (see above)
+python -m fedmlb.dataset_preparation dataset_config.alpha_dirichlet=0.3 total_clients=500
+
+# this will generate the setting for 3. (see above)
+python -m fedmlb.dataset_preparation dataset_config.alpha_dirichlet=0.6 
+
+# this will generate the setting for 4. (see above)
+python -m fedmlb.dataset_preparation dataset_config.alpha_dirichlet=0.6 total_clients=500
+```
+Note that, to reproduce those settings, we leverage the `.txt` files
+contained in the `client_data` folder in this project. Such files store
```

Comment:
Any chance we could remove the `client_data` files from the directory ? (they are ~7MB in total). Is there an obvious way of constructing those files via a not-too-complex script? -- we can naturally request people to git-clone them from the original repo you mention below, but that might not be always reliable.

Reply:
Those files could be constructed via a script by using a Dirichlet distribution just as the original paper. However, for reproducibility puproses, I decided to exactly compose the clients' dataset as they were crafted in the original paper. The files you mention, in fact, contain the IDs of the images assigned to that client, leveraging a Dirichlet distribution with a specific concentration parameters to select a certain number of images for a certain label. Obviously, if you re-run such a script, you could not be able to reproduce that dataset compositions unless you know the seed used to set the pseudo-random generation of numbers. So, in principle, I can produce a script that generates datasets that follows a Dirichlet distribution of labels among clients with a certain concentration parameter (but datasets would be different each run), or I can find a better way of store the data contained in the files under `client_data`.

For now, I've deleted some unused `.txt` files from the folder.

Reply:
Ok. I'll think about this and discuss with the others in the team. Now the files are 4MB (down from 7MB) so that's nice to see. Th
ere are other baselines that also have some not-so-small files as part of their proposed PR, so I'll update this thread once i figure out what's the best way to deal with these. Maybe keeping them is fine. Let's see...